### DISCUSSION TITLE
Validation: combination of required_if with prohibited_unless

### DISCUSSION THREAD

**Discussion by firebed** [laravel/framework#52370] [Ideas]
I have come across a lot of times in my projects where a field under validation is required if another field has a specific value, and the same field under validation is prohibited unless that condition is met.

For example I have an eshop and I want to adjust a product's stock with a specific adjustment reason.
The adjustment reason might be something like `1 => Damaged`, `2 => Theft or loss`, `3 => Promotion`, `99 => Other`.
If the reason code is `99` I expect the user to provide the other adjustment reason title, otherwise the user should be prohibited from providing a custom reason title.

Currently, alongside with custom rule, another possible way to achieve this is to use a combination of required_if and prohibited_unless. This is a part of my rules:

```php
'quantity' => ['required', 'integer'],
'adjustment_reason' => ['required', 'in:1,2,3,99'],
'other_adjustment_reason_title' => ['nullable', 'required_if:adjustment_reason,99', 'prohibited_unless:adjustment_reason,99'],
```

The above will yield one of the following errors based on the `adjustment_reason`

```php
[
    'adjustment_reason' => 99,
    'other_adjustment_reason_title' => ''
];

Error: `The other_adjustment_reason_title field is required when adjustment_reason is 99.`
```

```php
[
    'adjustment_reason' => 1,
    'other_adjustment_reason_title' => 'Return retail sale'
];

Error: `The other_adjustment_reason_title field is prohibited when adjustment_reason is 1.`
```

This is totally ok and is working. However I have a lot of such situations and I end up repeating myself all over the place which makes the project harder to maintain.

So, my solution to this is a new validation rule `exclusive_*`:
```php
Before:
'other_adjustment_reason_title' => ['nullable', 'required_if:adjustment_reason,99', 'prohibited_unless:adjustment_reason,99'],

After:
'other_adjustment_reason_title' => ['nullable', 'exclusive_if:adjustment_reason,99'],
```

The whole set: `exclusive_if`, `exclusive_unless`, `exclusive_with`, `exclusive_without`, `Rule::exclusiveIf()`.
I would love to make a PR on this but don't know if it is something worth working on. Any thoughts?

**Comment by shaedrich** [comment]
A similar case is with `required_with` + `prohibited_without` resp. `required_without` + `prohibited_with`. The rules have to be duplicated (reversed) for two fields each time. Sure, changing this, is a little more complicated, since the field names are keys, but maybe, there's also a solution

**Reply by firebed** [reply]
`required_with` + `prohibited_without` would be replaced by `exclusive_with`, meaning the field under validation is required when another field is present, and prohibited otherwise.
```php
'appointment_date' => ['required', 'date'],
'appointment_place' => ['nullable', 'exclusive_with:appointment_date'],
```
If user provides `appointment_date` he must also provide `appointment_place`.
If user doesn't provide `appointment_date` he must also not provide `appointment_place`.

Same with `required_without` + `prohibited_with` would be replaced by `exclusive_without`, meaning the field under validation is required when another field is not present, and prhibited if the another field is present.
```php
'primary_contact' => ['nullable', 'string'],
'alternate_contact' => ['nullable', 'exclusive_without:primary_contact'],
```
If user provides `primary_contact` he must not provide `alternate_contact`.
If user doesn't provide `primary_contact` he must provide `alternate_contact`.

**Reply by shaedrich** [reply]
Yep, sounds good 👍🏻
